include/seqan3/alignment/matrix/edit_distance_trace_matrix_full.hpp

```diff
@@ -216,4 +234,140 @@ class edit_distance_trace_matrix_full
+ * these values (normally seqan3::detail::trace_directions can be a combination of these values):
+ * * seqan3::detail::trace_directions::left
+ * * seqan3::detail::trace_directions::up
+ * * seqan3::detail::trace_directions::diagonal
+ *
+ * This requirement is needed to use the seqan3::detail::aligned_sequence_builder.
+ * \extends std::input_iterator
+ */
+template <typename word_t, bool is_semi_global, bool use_max_errors>
+struct edit_distance_trace_matrix_full<word_t, is_semi_global, use_max_errors>::trace_path_iterator
+{
+    /*!\name Associated types
+     * \{
+     */
+    //!\brief Input iterator tag.
+    using iterator_category = std::input_iterator_tag;
+    //!\copydoc seqan3::detail::trace_iterator_base::value_type
+    using value_type = detail::trace_directions;
+    //!\copydoc seqan3::detail::trace_iterator_base::difference_type
+    using difference_type = std::ptrdiff_t;
+    //!\}
+
+    //!\brief Shortcut for seqan3::detail::trace_directions::diagonal
+    constexpr static value_type D = value_type::diagonal;
+    //!\brief Shortcut for seqan3::detail::trace_directions::left
+    constexpr static value_type L = value_type::left;
+    //!\brief Shortcut for seqan3::detail::trace_directions::up
+    constexpr static value_type U = value_type::up;
+    //!\brief Shortcut for seqan3::detail::trace_directions::none
+    constexpr static value_type N = value_type::none;
+
+    /*!\name Element access
+     * \{
+     */
+    //!\copydoc seqan3::detail::trace_iterator_base::operator*
+    constexpr value_type operator*() const
+    {
+        value_type dir = parent->at(coordinate());
+
+        if (dir == N)
+            return N;
+
+        if ((dir & L) == L)
+            return L;
+        else if ((dir & U) == U)
+            return U;
+        else
+            return D;
+    }
+
+    //!\copydoc seqan3::detail::trace_iterator_base::coordinate
+    [[nodiscard]] constexpr matrix_coordinate const & coordinate() const
+    {
+        return coordinate_;
+    }
+    //!\}
+
+    /*!\name Arithmetic operators
+     * \{
+     */
+    //!\copydoc seqan3::detail::trace_iterator_base::operator++
+    constexpr trace_path_iterator & operator++()
+    {
+        value_type dir = *(*this);
+
+        if ((dir & L) == L)
+        {
+            coordinate_.col = std::max<size_t>(coordinate_.col, 1) - 1;
+        }
+        else if ((dir & U) == U)
+        {
+            coordinate_.row = std::max<size_t>(coordinate_.row, 1) - 1;
+        }
+        else if ((dir & D) == D)
+        {
+            coordinate_.row = std::max<size_t>(coordinate_.row, 1) - 1;
+            coordinate_.col = std::max<size_t>(coordinate_.col, 1) - 1;
+        }
+        else
+        {
+#ifndef NDEBUG
+            if (!(coordinate_.row == 0 || coordinate_.col == 0))
+                throw std::logic_error{"Unknown seqan3::trace_direction in an inner cell of the trace matrix."};
```

Comment:
coverage showed that I don't test this line (it wasn't tested before, too)

Reply:
Don't we usually use `assert` for these types of errors instead a `ifdef`?
And what happens if I step into this else clause in release mode, since I do not move the coordinate row and col, wouldn't it cause an infinite loop in this unknown trace cell?

Reply:
true we could just assert here, I just copied the old code over.